File: 백준/Gold/14567. 선수과목 （Prerequisite）/선수과목 （Prerequisite）.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        
        List<Integer>[] pre = new ArrayList[n + 1];
        int[] wait = new int[n + 1];
        int[] result = new int[n + 1];
        
        for (int i = 0; i <= n; i++) {
            pre[i] = new ArrayList<>();
        }
        
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            pre[a].add(b);
            wait[b]++;
        }
        
        Queue<Integer> work = new LinkedList<>();
        for (int i = 1; i <= n; i++) {
            if (wait[i] == 0) {
                work.add(i);
            }
        }
        
        int time = 0;
        while (!work.isEmpty()) {
            time++;
            Queue<Integer> nextWork = new LinkedList<>();
            while (!work.isEmpty()) {
                int study = work.poll();
                result[study] = time;
                for (int next : pre[study]) {
                    wait[next]--;
                    if (wait[next] == 0) {
                        nextWork.add(next);
                    }
                }
            }
            work = nextWork;
        }
        
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= n; i++) {
            sb.append(result[i]).append(" ");
        }
        System.out.println(sb.toString().trim());
    }
}

File: 백준/Gold/14567. 선수과목 （Prerequisite）/선수과목 （Prerequisite）.py
import sys
input = lambda : sys.stdin.readline().rstrip()

n,m = map(int,input().split())
pre = [[] for i in range(n+1)]
wait = [0]*(n+1)
wait[0] = -1
result = [0]*(n+1)
for _ in range(m):
    a,b = map(int,input().split())
    pre[a].append(b)
    wait[b]+=1

work=[i for i in range(1,n+1) if wait[i]==0]
time = 0
while True:
    time+=1
    next_work=[]
    # print('current time',time)
    # print('current work',work)
    while work:
        study = work.pop()
        # print('pop',study)
        result[study] = time
        for next in pre[study]:
            wait[next]-=1
            if wait[next]==0:
                next_work.append(next)
    work= next_work
    if not work:
        break
print(*result[1:])
        
    
